pennylane/qaoa/mixers.py

```diff
@@ -112,3 +89,73 @@ def xy_mixer(graph):
         obs.append(qml.PauliY(node1) @ qml.PauliY(node2))
 
     return qml.Hamiltonian(coeffs, obs)
+
+
+def bit_flip_mixer(graph, n):
+    r"""Creates a bit-flip mixer Hamiltonian.
+    This mixer is defined as:
+    .. math:: H_M \ = \ \displaystyle\sum_{v \in V(G)} \frac{1}{2^d(v)} X_{v}
+              \displaystyle\prod_{w \in \text{nbhd}(v)} (\mathbb{I} \ + \ (-1)^n Z_w)
+    where :math:`V(G)` is the set of vertices of some graph :math:`G`, :math:`d(v)` is the
+    `degree <https://en.wikipedia.org/wiki/Degree_(graph_theory)>`__ of vertex :math:`v`, and
+    :math:`\text{nbhd}(v)` is the `neighbourhood <https://en.wikipedia.org/wiki/Neighbourhood_(graph_theory)>`__
+    of vertex :math:`v`. In addition, :math:`\mathbb{I}` is the identity operator, and :math:`X_v` and :math:`Z_v`
+    are the Pauli-Z and Pauli-X operators on vertex :math:`v`, respectively.
+    This mixer was introduced in *From the Quantum Approximate Optimization Algorithm
+    to a Quantum Alternating Operator Ansatz* by Stuart Hadfield, Zhihui Wang, Bryan O'Gorman,
+    Eleanor G. Rieffel, Davide Venturelli, and Rupak Biswas [`arXiv:1709.03489 <https://arxiv.org/abs/1709.03489>`__].
+    Args:
+         graph (nx.Graph): A graph defining the collections of wires on which the Hamiltonian acts
+         n (int): Either :math:`0` or :math:`1`. Determines whether the mixer performs a conditional
+                  bit-flip of a vertex if all neighbouring vertices are labelled :math:`0` or :math:`1`.
+    Returns:
+        Hamiltonian: Mixer Hamiltonian
+    .. UsageDetails::
+        The mixer Hamiltonian can be called as follows:
+        .. code-block:: python3
+            from pennylane import qaoa
+            from networkx import Graph
+            graph = Graph([(0, 1), (1, 2)])
+            mixer_h = qaoa.bit_flip_mixer(graph, 0)
+        >>> print(mixer_h)
+        (0.5) [X0]
+        + (0.5) [X0 Z1]
+        + (0.25) [X1]
+        + (0.25) [X1 Z2]
+        + (0.25) [X1 Z0]
+        + (0.25) [X1 Z0 Z2]
+        + (0.5) [X2]
+        + (0.5) [X2 Z1]
+    """
+
+    if not isinstance(graph, nx.Graph):
+        raise ValueError(
+            "Input graph must be a nx.Graph object, got {}".format(type(graph).__name__)
+        )
+
+    if n not in [0, 1]:
+        raise ValueError("'n' must be either 0 or 1, got {}".format(n))
+
+    sign = 1 if n == 0 else -1
+
+    coeffs = []
+    terms = []
+
+    for i in graph.nodes:
+
+        neighbours = list(graph.neighbors(i))
+        degree = len(neighbours)
+
+        n_terms = [[qml.PauliX(i)]] + [[qml.Identity(n), qml.PauliZ(n)] for n in neighbours]
+        n_coeffs = [[1, sign] for n in neighbours]
+
+        final_terms = [qml.operation.Tensor(*list(m)).prune() for m in itertools.product(*n_terms)]
+        final_coeffs = [
+            (0.5 ** degree) * functools.reduce(lambda x, y: x * y, list(m), 1)
+            for m in itertools.product(*n_coeffs)
+        ]
+
+        coeffs.extend(final_coeffs)
```

Comment:
Out of curiosity, why use `extend` instead of `append`?
Also, I don't think we need to initialize `coeffs` and `terms` as empty lists, but instead rename `final_terms` to `terms` and same with `final_coeffs` to `coeffs`

Reply:
We use extend because we wish to add each element in a list of elements to another list. I think we have to keep the empty lists, as we are cycling through multiple values of `i`. For each iteration, we append the outputted values to the lists that lie outside the loop.